src/adapter/adapterDiscovery.ts

```diff
@@ -0,0 +1,439 @@
+            // manufacturer: '',
+            pathRegex: '.*zigstar.*',
+        },
+    ],
+    zboss: [
+        {
+            // Nordic Zigbee NCP
+            vendorId: '2fe3',
+            productId: '0100',
+            manufacturer: 'ZEPHYR',
+            // /dev/serial/by-id/usb-ZEPHYR_Zigbee_NCP_54ACCFAFA6DAD111-if00
+            pathRegex: '.*ZEPHYR.*',
+        },
+    ],
+    zigate: [
+        {
+            // ZiGate PL2303HX (blue)
+            vendorId: '067b',
+            productId: '2303',
+            manufacturer: 'zigate_PL2303',
+            pathRegex: '.*zigate.*',
+        },
+        {
+            // ZiGate CP2102 (red)
+            vendorId: '10c4',
+            productId: 'ea60',
+            manufacturer: 'zigate_cp2102',
+            pathRegex: '.*zigate.*',
+        },
+        {
+            // ZiGate+ V2 CDM_21228
+            vendorId: '0403',
+            productId: '6015',
+            // manufacturer: '',
+            // /dev/serial/by-id/usb-FTDI_ZiGate_ZIGATE+-if00-port0
+            pathRegex: '.*zigate.*',
+        },
+    ],
+};
+
+async function getSerialPortList(): Promise<PortInfo[]> {
+    const portInfos = await SerialPort.list();
+
+    // TODO: can sorting be removed in favor of `path` regex matching?
+
+    // CC1352P_2 and CC26X2R1 lists as 2 USB devices with same manufacturer, productId and vendorId
+    // one is the actual chip interface, other is the XDS110.
+    // The chip is always exposed on the first one after alphabetical sorting.
+    /* istanbul ignore next */
+    portInfos.sort((a, b) => (a.path < b.path ? -1 : 1));
+
+    return portInfos;
+}
+
+/**
+ * Case insensitive string matching.
+ * @param str1
+ * @param str2
+ * @returns
+ */
+function matchString(str1: string, str2: string): boolean {
+    return str1.localeCompare(str2, undefined, {sensitivity: 'base'}) === 0;
+}
+
+/**
+ * Case insensitive regex matching.
+ * @param regexStr Passed to RegExp constructor.
+ * @param str Always returns false if undefined.
+ * @returns
+ */
+function matchRegex(regexStr: string, str?: string): boolean {
+    return str !== undefined && new RegExp(regexStr, 'i').test(str);
+}
+
+function matchUSBFingerprint(
+    portInfo: PortInfo,
+    isWindows: boolean,
+    entries: USBAdapterFingerprint[],
+): [PortInfo['path'], USBAdapterFingerprint] | undefined {
+    if (!portInfo.vendorId || !portInfo.productId) {
+        // port info is missing essential information for proper matching, ignore it
+        return;
+    }
+
+    for (const entry of entries) {
+        if (
+            matchString(portInfo.vendorId, entry.vendorId) &&
+            matchString(portInfo.productId, entry.productId) &&
+            (!entry.manufacturer || !portInfo.manufacturer || matchString(portInfo.manufacturer, entry.manufacturer) || isWindows) &&
+            (!entry.pathRegex || matchRegex(entry.pathRegex, portInfo.path) || matchRegex(entry.pathRegex, portInfo.pnpId) || isWindows)
```

Comment:
Currently both the vendorId, productId and path should match, however I would argue that a match on either vendorId + productId OR path should be enough:
- Vendor + productID might be unknown (e.g. insufficient right for udev), then we can still match on the path
- Vendor + productID is known, but the path is not a `serial-by-id`, 2 cases I see here:
  - User uses e.g. `/dev/ttyUSB0`
  - Customised device name with Docker when running with `--device /dev/serial/by-id/usb-Silicon_Labs_CP2102N_USB_to_UART_Bridge_Controller_de4374fdea93eb11b514a85b3d98b6d1-if00-port0:/dev/adapter` (which is what I do)
  -

Reply:
It would, but it wouldn't work... Cases where path might be the same (except for id hex) would match (like some Dongle-E vs some Dongle-P, only way to differentiate is with vendorId/productId).

Reply:
> (like some Dongle-E vs some Dongle-P, only way to differentiate is with vendorId/productId).

Dongle-E switched to cp2102 due to high failure rates in ch9102. The initial batches with cp2102 were flashed with the Dongle-P usb config, so are indistinguishable from Dongle-P. I believe this has been corrected now with latest devices, but it leaves many devices that can only be discovered by probing firmware  :(

Reply:
Yes USB discovery definitely has its limits. Between mistakes, lack of details, various versions... We really can only "try to match as many as possible", the rest requires manual configuration.